programs/stake/src/stake_state.rs

```diff
@@ -1030,49 +1042,50 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
 
     fn merge(
         &self,
-        source_stake: &KeyedAccount,
+        source_account: &KeyedAccount,
         clock: &Clock,
         stake_history: &StakeHistory,
         signers: &HashSet<Pubkey>,
     ) -> Result<(), InstructionError> {
-        if source_stake.owner()? != id() {
+        if source_account.owner()? != id() {
             return Err(InstructionError::IncorrectProgramId);
         }
 
-        let meta = match self.state()? {
-            StakeState::Stake(meta, stake) => {
-                // stake must be fully de-activated
-                if stake.stake(clock.epoch, Some(stake_history), true) != 0 {
-                    return Err(StakeError::MergeActivatedStake.into());
-                }
-                meta
-            }
-            StakeState::Initialized(meta) => meta,
-            _ => return Err(InstructionError::InvalidAccountData),
-        };
+        let (meta, stake) = get_info_if_mergeable(self, clock, stake_history)?;
+
         // Authorized staker is allowed to split/merge accounts
         meta.authorized.check(signers, StakeAuthorize::Staker)?;
 
-        let source_meta = match source_stake.state()? {
-            StakeState::Stake(meta, stake) => {
-                // stake must be fully de-activated
-                if stake.stake(clock.epoch, Some(stake_history), true) != 0 {
-                    return Err(StakeError::MergeActivatedStake.into());
-                }
-                meta
-            }
-            StakeState::Initialized(meta) => meta,
-            _ => return Err(InstructionError::InvalidAccountData),
-        };
+        let (source_meta, source_stake) =
+            get_info_if_mergeable(source_account, clock, stake_history)?;
 
         // Meta must match for both accounts
         if meta != source_meta {
             return Err(StakeError::MergeMismatch.into());
         }
 
+        // If both stakes are fully-active and their delegations are
+        // compatiible, merge the delegations. Do NOT merge mismatched
+        // delegated and deactivated stakes!
+        match (stake, source_stake) {
+            (Some(mut stake), Some(mut source_stake)) => {
+                if stake.can_merge_with(&source_stake) {
+                    let stake_lamports = source_stake.delegation.stake;
+                    source_stake.delegation.stake -= stake_lamports;
+                    stake.delegation.stake += stake_lamports;
```

Comment:
So this means that on merge, the source rent_exempt_reserve just becomes excess (non-stake) lamports in the merge account, right? That feels a little weird, but I guess that's the best option.